clippy_lints/src/hashset_insert_after_contains.rs

```diff
@@ -0,0 +1,184 @@
+            ..
+        }) = higher::If::hir(expr)
+        else {
+            return;
+        };
+
+        let Some(contains_expr) = try_parse_contains(cx, cond_expr) else {
+            return;
+        };
+
+        if !find_insert_calls(cx, &contains_expr, then_expr) {
+            return;
+        };
+        span_lint(
+            cx,
+            HASHSET_INSERT_AFTER_CONTAINS,
+            expr.span,
+            "usage of `HashSet::insert` after `HashSet::contains`. Remove the usage of `HashSet::contains` and just call `HashSet::contains` instead",
+        );
+    }
+}
+
+struct ContainsExpr<'tcx> {
+    receiver: &'tcx Expr<'tcx>,
+    value: &'tcx Expr<'tcx>,
+}
+fn try_parse_contains<'tcx>(cx: &LateContext<'_>, expr: &'tcx Expr<'_>) -> Option<ContainsExpr<'tcx>> {
+    let expr = peel_hir_expr_while(expr, |e| match e.kind {
+        ExprKind::Unary(UnOp::Not, e) => Some(e),
+        _ => None,
+    });
+    match expr.kind {
+        ExprKind::MethodCall(
+            path,
+            receiver,
+            [
+                Expr {
+                    kind: ExprKind::AddrOf(_, _, value),
+                    span: value_span,
+                    ..
+                },
+            ],
+            _,
+        ) => {
+            let receiver_ty = cx.typeck_results().expr_ty(receiver);
+            if value_span.ctxt() == expr.span.ctxt()
+                && is_type_diagnostic_item(cx, receiver_ty, sym::HashSet)
+                && path.ident.name == sym!(contains)
+            {
+                Some(ContainsExpr { receiver, value })
+            } else {
+                None
+            }
+        },
+        _ => None,
+    }
+}
+
+struct InsertExpr<'tcx> {
+    receiver: &'tcx Expr<'tcx>,
+    value: &'tcx Expr<'tcx>,
+}
+fn try_parse_insert<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) -> Option<InsertExpr<'tcx>> {
+    if let ExprKind::MethodCall(path, receiver, [value], _) = expr.kind {
+        let receiver_ty = cx.typeck_results().expr_ty(receiver);
+        if is_type_diagnostic_item(cx, receiver_ty, sym::HashSet) && path.ident.name == sym!(insert) {
+            Some(InsertExpr { receiver, value })
+        } else {
+            None
+        }
+    } else {
+        None
+    }
+}
+
+fn find_insert_calls<'tcx>(cx: &LateContext<'tcx>, contains_expr: &ContainsExpr<'tcx>, expr: &'tcx Expr<'_>) -> bool {
+    let mut s = InsertSearcher {
+        cx,
+        receiver: contains_expr.receiver,
+        value: contains_expr.value,
+        should_lint: false,
+    };
+    s.visit_expr(expr);
+    s.should_lint
+}
+
+struct InsertSearcher<'cx, 'tcx> {
+    cx: &'cx LateContext<'tcx>,
+    /// The receiver expression used in the contains call.
+    receiver: &'tcx Expr<'tcx>,
+    /// The value expression used in the contains call.
+    value: &'tcx Expr<'tcx>,
+    /// Whether or a lint shoud be emitted.
+    should_lint: bool,
+}
+
+impl<'tcx> Visitor<'tcx> for InsertSearcher<'_, 'tcx> {
```

Comment:
I think, instead of using a visitor, this could use `for_each_expr` on the `then` block. I *think* you could copy the code from `visit_expr` into that closure, and change `Some` to `Break`.